File: 2025/01 January/08 Count the number of possible triangles/main.py
# Link: https://www.geeksforgeeks.org/problems/count-possible-triangles-1587115620/1


class Solution:
    def countTriangles(self, arr: list[int]) -> int:
        n = len(arr)
        if n < 3:
            return 0

        arr.sort()
        triangleCount = 0
        for k in range(n - 1, 1, -1):
            i, j = 0, k - 1
            while i < j:
                if arr[i] + arr[j] > arr[k]:
                    triangleCount += j - i
                    j -= 1
                else:
                    i += 1
        return triangleCount


if __name__ == "__main__":
    solution = Solution()
    arr = [4, 6, 3, 7]
    print(solution.countTriangles(arr))

    arr = [10, 21, 22, 100, 101, 200, 300]
    print(solution.countTriangles(arr))

    arr = [1, 2, 3]
    print(solution.countTriangles(arr))

File: 2025/01 January/08 Count the number of possible triangles/main.cpp
// Link: https://www.geeksforgeeks.org/problems/count-possible-triangles-1587115620/1

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int countTriangles(vector<int> &arr)
    {
        int n = arr.size();
        if (n < 3)
            return 0;
        ;
        sort(arr.begin(), arr.end());
        int triangleCount = 0;
        for (int k = n - 1; k >= 2; k--)
        {
            int i = 0, j = k - 1;
            while (i < j)
            {
                if (arr[i] + arr[j] > arr[k])
                {
                    triangleCount += (j - i);
                    j--;
                }
                else
                {
                    i++;
                }
            }
        }
        return triangleCount;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {4, 6, 3, 7};
    cout << object.countTriangles(arr) << endl;

    arr = {10, 21, 22, 100, 101, 200, 300};
    cout << object.countTriangles(arr) << endl;

    arr = {1, 2, 3};
    cout << object.countTriangles(arr) << endl;
    return 0;
}